### PR TITLE
feat: adds Collection.key

### PR DESCRIPTION
**Pull request by pomegranited**
## Description

Adds a new required field to the Collection model: `key` (string), which uniquely identifies a Collection within a Learning Package.
Existing Collection keys are initialized with a random string.

## Supporting information

Part of : https://github.com/openedx/modular-learning/issues/226

Private-ref: [FAL-3783](https://tasks.opencraft.com/browse/FAL-3783)

## Testing instructions

See https://github.com/openedx/edx-platform/pull/35321 for full testing instructions.

Testing migrations:

In an openedx-learning virtual environment, run:

0. Create an openedx-learning virtual environment
1. Install openedx-learning @ `main`, and run `python3 manage.py migrate`
2. Create a collection `python3 manage.py shell`:
   ```
   import openedx_learning.api.authoring as api
   learning_package = api.create_learning_package("LP1", "Learning Package 1")
   for x in range(0, 11):
       api.create_collection(learning_package.id, title=f"Collection {x}", created_by=None)
       # <Collection> (1:Collection 1)
       # <Collection> (2:Collection 2)
       ...
       # <Collection> (10:Collection 10)
   ```
2. Install openedx-learning @ `open-craft/jill/collection-key` 
3. Run `python3 manage.py migrate`
4. Check that a key was generated for your collection  `python3 manage.py shell`:
   ```
   import openedx_learning.api.authoring as api
   learning_package = api.get_learning_package_by_key("LP1")
   collections = api.get_collections(learning_package.id)
   for collection in collections.all():
       print(collection.key)
       # gy19KxxV9xrjEMmR4TRNeE2p8sEbQfetVjZPzKPUbHURAgAgJY
       # RNeE2p8sEbQfetVjZPzKPUbgy19KxxV9xrjEMmR4TRNeE2p8sEb
       ...
   ```

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @pomegranited!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.


#### :radio_button: Update the status of your PR

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.




#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@axim-engineering`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

**Review by rpenido — CHANGES_REQUESTED** [review]
Hi @pomegranited! This is looking good!

Only a few nits

File: openedx_learning/apps/authoring/collections/models.py [hunk 1]
```diff
@@ -153,6 +173,7 @@ class Meta:
         verbose_name_plural = "Collections"
         indexes = [
             models.Index(fields=["learning_package", "title"]),
+            models.Index(fields=["learning_package", "key"]),
         ]
```
[the remark this answers is not in the document]
Comment (pomegranited) [reply openedx_learning/apps/authoring/collections/models.py:177]: Good catch, thank you!

File: openedx_learning/apps/authoring/collections/models.py [hunk 2]
```diff
@@ -165,7 +186,7 @@ def __str__(self) -> str:
         """
         User-facing string representation of a Collection.
         """
-        return f"<{self.__class__.__name__}> ({self.id}:{self.title})"
+        return f"<{self.__class__.__name__}> ({self.key}:{self.title})"
```
[the remark this answers is not in the document]
Comment (pomegranited) [reply openedx_learning/apps/authoring/collections/models.py:189]: Sure, can do.

File: openedx_learning/apps/authoring/collections/migrations/0004_collection_key.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        ('oel_collections', '0003_collection_entities'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='collection',
+            name='key',
+            field=openedx_learning.lib.fields.MultiCollationCharField(
+                db_collations={'mysql': 'utf8mb4_bin', 'sqlite': 'BINARY'}, db_column='_key', max_length=500,
+                default=generate_key,
```
[the remark this answers is not in the document]
Comment (pomegranited) [reply openedx_learning/apps/authoring/collections/migrations/0004_collection_key.py]: Ach.. sorry, I should have made a proper data migration. Fixed with [8b75af1](https://github.com/openedx/openedx-learning/pull/223/commits/8b75af1965bf8179401cfcf262b6cd988bce4125), and updated the PR instructions to test it.

File: openedx_learning/apps/authoring/collections/api.py [hunk 4]
```diff
@@ -48,22 +51,24 @@ def create_collection(
...
+    return Collection.objects.get_by_key(learning_package_id, collection_key)
 
 
 def update_collection(
-    collection_id: int,
+    learning_package_id: int,
+    key: str,
+    *,
     title: str | None = None,
     description: str | None = None,
```
Comment (rpenido) [inline openedx_learning/apps/authoring/collections/api.py:66]: Sorry for not catching this before. :disappointed: 
I think we should be able to update the key here.
```suggestion
    description: str | None = None,
    new_key: str
```

**Review by rpenido — COMMENTED** [review]
LGTM :+1:
Thank you for your work, @pomegranited!

I commented about adding a new parameter to allow the user to update the key. Other than that, everything is good on my end.

- [x] I tested this using the instructions from https://github.com/openedx/edx-platform/pull/35321
- [x] I read through the code
- [x] ~I checked for accessibility issues~
- [x] Includes documentation

File: openedx_learning/apps/authoring/collections/api.py (hunk 4 above)
Comment (pomegranited) [reply openedx_learning/apps/authoring/collections/api.py:66]: Hmm.. if we allow that, it'll change the [Collection's opaque key](https://github.com/openedx/opaque-keys/pull/335), which will disconnect it from any applied tags.. so I think it should be immutable.

Comment (rpenido) [reply openedx_learning/apps/authoring/collections/api.py:66]: You're right! My bad!

**Comment by pomegranited** [comment]
Hi @ormsbee , this is a minor PR, but I'd appreciate your review if you have time?

**Review by rpenido — APPROVED** [review]
Thank you for the changes, @pomegranited!

LGTM :+1:

**Pull request merged**